File: rust/libipc/tests/test_channel_stress.rs
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya contributors
//
// Port of high-throughput prod_cons tests from cpp-ipc/test/archive/test_queue.cpp.
// Stress tests for the channel with varying sender/receiver counts and high message volumes.

use std::sync::atomic::{AtomicI32, AtomicU64, Ordering};
use std::sync::Arc;
use std::thread;
use std::time::{Duration, Instant};

use libipc::{Channel, Mode, Route};

static COUNTER: AtomicI32 = AtomicI32::new(0);

fn unique_name(tag: &str) -> String {
    let n = COUNTER.fetch_add(1, Ordering::Relaxed);
    format!("stress_{tag}_{n}_{}", std::process::id())
}

// Port of Queue.prod_cons_1v1 — single sender, single receiver, many messages
#[test]
fn route_1v1_throughput() {
    let name = unique_name("r_1v1");
    Route::clear_storage(&name);

    let msg_count = 1000usize;
    let received = Arc::new(AtomicU64::new(0));

    let n = name.clone();
    let rc = Arc::clone(&received);
    let receiver = thread::spawn(move || {
        let mut r = Route::connect(&n, Mode::Receiver).expect("receiver");
        for _ in 0..msg_count {
            let buf = r.recv(Some(5000)).expect("recv");
            if !buf.is_empty() {
                rc.fetch_add(1, Ordering::Relaxed);
            }
        }
    });

    thread::sleep(Duration::from_millis(50));

    let mut sender = Route::connect(&name, Mode::Sender).expect("sender");
    sender.wait_for_recv(1, Some(2000)).expect("wait");

    let start = Instant::now();
    for i in 0..msg_count {
        let msg = i.to_le_bytes();
        assert!(sender.send(&msg, 5000).expect("send"));
    }
    let elapsed = start.elapsed();

    receiver.join().unwrap();

    assert_eq!(received.load(Ordering::Relaxed), msg_count as u64);
    eprintln!(
        "route 1v1: {msg_count} msgs in {:.1}ms ({:.0} msg/s)",
        elapsed.as_secs_f64() * 1000.0,
        msg_count as f64 / elapsed.as_secs_f64()
    );
}

// Port of Queue.prod_cons_1vN_broadcast — 1 sender, N receivers
#[test]
fn route_1vn_broadcast() {
    for num_receivers in [2, 4] {
        let name = unique_name("r_1vn");
        Route::clear_storage(&name);

        let msg_count = 500usize;
        let total_received = Arc::new(AtomicU64::new(0));

        let mut receivers = Vec::new();
        for _ in 0..num_receivers {
            let n = name.clone();
            let rc = Arc::clone(&total_received);
            receivers.push(thread::spawn(move || {
                let mut r = Route::connect(&n, Mode::Receiver).expect("receiver");
                for _ in 0..msg_count {
                    let buf = r.recv(Some(5000)).expect("recv");
                    if !buf.is_empty() {
                        rc.fetch_add(1, Ordering::Relaxed);
                    }
                }
            }));
        }

        thread::sleep(Duration::from_millis(100));

        let mut sender = Route::connect(&name, Mode::Sender).expect("sender");
        sender
            .wait_for_recv(num_receivers, Some(2000))
            .expect("wait");

        let start = Instant::now();
        for i in 0..msg_count {
            let msg = i.to_le_bytes();
            assert!(sender.send(&msg, 5000).expect("send"));
        }
        let elapsed = start.elapsed();

        for r in receivers {
            r.join().unwrap();
        }

        let total = total_received.load(Ordering::Relaxed);
        assert_eq!(total, (msg_count * num_receivers) as u64);
        eprintln!(
            "route 1v{num_receivers}: {msg_count} msgs in {:.1}ms ({:.0} msg/s, {total} total recvd)",
            elapsed.as_secs_f64() * 1000.0,
            msg_count as f64 / elapsed.as_secs_f64()
        );
    }
}

// Port of Queue.prod_cons_NvN_broadcast — N senders, N receivers
#[test]
fn channel_nvn_broadcast() {
    for n in [2, 3] {
        let name = unique_name("c_nvn");
        Channel::clear_storage(&name);

        let msg_per_sender = 100usize;
        let total_msgs = n * msg_per_sender;
        let total_sent = Arc::new(AtomicU64::new(0));
        let total_received = Arc::new(AtomicU64::new(0));

        let mut receivers = Vec::new();
        for _ in 0..n {
            let nm = name.clone();
            let rc = Arc::clone(&total_received);
            receivers.push(thread::spawn(move || {
                let mut ch = Channel::connect(&nm, Mode::Receiver).expect("receiver");
                for _ in 0..total_msgs {
                    let buf = ch.recv(Some(5000)).expect("recv");
                    if !buf.is_empty() {
                        rc.fetch_add(1, Ordering::Relaxed);
                    }
                }
            }));
        }

        thread::sleep(Duration::from_millis(200));

        let mut senders = Vec::new();
        for s in 0..n {
            let nm = name.clone();
            let sc = Arc::clone(&total_sent);
            senders.push(thread::spawn(move || {
                let mut ch = Channel::connect(&nm, Mode::Sender).expect("sender");
                ch.wait_for_recv(n, Some(3000)).expect("wait");
                for j in 0..msg_per_sender {
                    let msg = format!("S{s}M{j}");
                    if ch.send(msg.as_bytes(), 5000).expect("send") {
                        sc.fetch_add(1, Ordering::Relaxed);
                    }
                }
            }));
        }

        for s in senders {
            s.join().unwrap();
        }
        for r in receivers {
            r.join().unwrap();
        }

        let sent = total_sent.load(Ordering::Relaxed);
        let received = total_received.load(Ordering::Relaxed);
        assert_eq!(sent, total_msgs as u64);
        assert_eq!(received, (total_msgs * n) as u64);
        eprintln!("channel {n}v{n}: {total_msgs} msgs, sent={sent}, received={received}");
    }
}

// Port of IPC.Nv1 — N senders, 1 receiver broadcast
#[test]
fn channel_nv1_broadcast() {
    for num_senders in [2usize, 4] {
        let name = unique_name("c_nv1");
        Channel::clear_storage(&name);

        let msg_per_sender = 100usize;
        let total_msgs = num_senders * msg_per_sender;
        let total_received = Arc::new(AtomicU64::new(0));

        let nm = name.clone();
        let rc = Arc::clone(&total_received);
        let receiver = thread::spawn(move || {
            let mut ch = Channel::connect(&nm, Mode::Receiver).expect("receiver");
            for _ in 0..total_msgs {
                let buf = ch.recv(Some(5000)).expect("recv");
                if !buf.is_empty() {
                    rc.fetch_add(1, Ordering::Relaxed);
                }
            }
        });

        thread::sleep(Duration::from_millis(100));

        let total_sent = Arc::new(AtomicU64::new(0));
        let mut senders = Vec::new();
        for s in 0..num_senders {
            let nm = name.clone();
            let sc = Arc::clone(&total_sent);
            senders.push(thread::spawn(move || {
                let mut ch = Channel::connect(&nm, Mode::Sender).expect("sender");
                ch.wait_for_recv(1, Some(3000)).expect("wait");
                for j in 0..msg_per_sender {
                    let msg = format!("S{s}M{j}");
                    if ch.send(msg.as_bytes(), 5000).expect("send") {
                        sc.fetch_add(1, Ordering::Relaxed);
                    }
                }
            }));
        }

        for s in senders {
            s.join().unwrap();
        }
        receiver.join().unwrap();

        let sent = total_sent.load(Ordering::Relaxed);
        let received = total_received.load(Ordering::Relaxed);
        assert_eq!(sent, total_msgs as u64);
        assert_eq!(received, total_msgs as u64);
        eprintln!("channel {num_senders}v1: {total_msgs} msgs, sent={sent}, received={received}");
    }
}

// Stress: rapid connect/disconnect cycles
#[test]
fn channel_rapid_reconnect() {
    let name = unique_name("reconnect");
    Channel::clear_storage(&name);

    for i in 0..20 {
        let mut sender = Channel::connect(&name, Mode::Sender).expect("sender");
        let mut receiver = Channel::connect(&name, Mode::Receiver).expect("receiver");

        sender.wait_for_recv(1, Some(1000)).expect("wait");
        let msg = format!("iter{i}");
        assert!(sender.send(msg.as_bytes(), 1000).expect("send"));

        let buf = receiver.recv(Some(1000)).expect("recv");
        assert!(!buf.is_empty());
        assert_eq!(buf.data(), msg.as_bytes());
    }
}

// Stress: large messages with fragmentation under load
#[test]
fn route_large_messages_stress() {
    let name = unique_name("large_stress");
    Route::clear_storage(&name);

    let msg_count = 20usize;
    let msg_size = 1024usize; // 1KB — requires 16 ring slots per message

    let n = name.clone();
    let receiver = thread::spawn(move || {
        let mut r = Route::connect(&n, Mode::Receiver).expect("receiver");
        for i in 0..msg_count {
            let buf = r.recv(Some(10000)).expect("recv");
            assert_eq!(buf.len(), msg_size, "msg {i} wrong size");
            assert!(
                buf.data().iter().all(|&b| b == (i as u8)),
                "msg {i} corrupt"
            );
        }
    });

    thread::sleep(Duration::from_millis(50));

    let mut sender = Route::connect(&name, Mode::Sender).expect("sender");
    sender.wait_for_recv(1, Some(2000)).expect("wait");

    for i in 0..msg_count {
        let msg = vec![i as u8; msg_size];
        assert!(sender.send(&msg, 10000).expect("send"));
    }

    receiver.join().unwrap();
}

File: swift/libipc/Tests/LibIPCTests/TestChannelStress.swift
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya contributors
//
// Channel stress tests — port of rust/libipc/tests/test_channel_stress.rs

import Testing
@testable import LibIPC
import Darwin.POSIX

// MARK: - Tests

@Suite("Channel stress")
struct TestChannelStress {

    // Port of route_1v1_throughput — 1000 messages, verify all received
    @Test("route 1v1 — 1000 messages all received")
    func route1v1Throughput() async throws {
        let name = "swift_stress_r1v1_\(UInt32.random(in: 0..<UInt32.max))"
        defer { Task { await Route.clearStorage(name: name) } }

        let msgCount = 1000
        let sender = Route.connectBlocking(name: name, mode: .sender)
        let receiver = Route.connectBlocking(name: name, mode: .receiver)

        nonisolated(unsafe) var received = 0

        let recvThread = spawnPthread {
            for _ in 0..<msgCount {
                let buf = try! receiver.recv(timeout: .seconds(5))
                if !buf.isEmpty { received += 1 }
            }
        }

        _ = try sender.waitForRecv(count: 1, timeout: .seconds(2))
        for i in 0..<msgCount {
            var bytes = withUnsafeBytes(of: i.littleEndian) { Array($0) }
            _ = try sender.send(data: bytes, timeout: .seconds(5))
        }

        await joinThread(recvThread)
        sender.disconnect(); receiver.disconnect()
        #expect(received == msgCount)
    }

    // Port of route_1vn_broadcast — 1 sender, N receivers, each gets all messages
    @Test("route 1vN broadcast — 2 and 4 receivers")
    func route1vNBroadcast() async throws {
        for numReceivers in [2, 4] {
            let name = "swift_stress_r1vn_\(UInt32.random(in: 0..<UInt32.max))"
            defer { Task { await Route.clearStorage(name: name) } }

            let msgCount = 200
            let sender = Route.connectBlocking(name: name, mode: .sender)
            let receivers = (0..<numReceivers).map { _ in
                Route.connectBlocking(name: name, mode: .receiver)
            }

            nonisolated(unsafe) var totalReceived = 0
            nonisolated(unsafe) var mu1 = pthread_mutex_t()
            pthread_mutex_init(&mu1, nil)

            let recvThreads = receivers.map { r in
                spawnPthread {
                    var got = 0
                    while got < msgCount {
                        let buf = try! r.recv(timeout: .seconds(5))
                        if !buf.isEmpty {
                            got += 1
                            pthread_mutex_lock(&mu1); totalReceived += 1; pthread_mutex_unlock(&mu1)
                        }
                    }
                }
            }

            _ = try sender.waitForRecv(count: numReceivers, timeout: .seconds(2))
            for i in 0..<msgCount {
                let bytes = withUnsafeBytes(of: i.littleEndian) { Array($0) }
                _ = try sender.send(data: bytes, timeout: .seconds(5))
            }

            for t in recvThreads { await joinThread(t) }
            pthread_mutex_destroy(&mu1)
            sender.disconnect()
            for r in receivers { r.disconnect() }
            #expect(totalReceived == msgCount * numReceivers)
        }
    }

    // Port of channel_nvn_broadcast — N senders, N receivers
    @Test("channel NvN broadcast — N=2 and N=3")
    func channelNvNBroadcast() async throws {
        for n in [2, 3] {
            let name = "swift_stress_cnvn_\(UInt32.random(in: 0..<UInt32.max))"
            defer { Task { await Channel.clearStorage(name: name) } }

            let msgPerSender = 100
            let totalMsgs = n * msgPerSender

            let ctrl = Channel.connectBlocking(name: name, mode: .sender)
            let receivers = (0..<n).map { _ in Channel.connectBlocking(name: name, mode: .receiver) }
            let senders   = (0..<n).map { _ in Channel.connectBlocking(name: name, mode: .sender) }

            nonisolated(unsafe) var totalSent     = 0
            nonisolated(unsafe) var totalReceived = 0
            nonisolated(unsafe) var mu2 = pthread_mutex_t()
            pthread_mutex_init(&mu2, nil)

            let recvThreads = receivers.map { ch in
                spawnPthread {
                    var got = 0
                    while got < totalMsgs {
                        let buf = try! ch.recv(timeout: .seconds(5))
                        if !buf.isEmpty {
                            got += 1
                            pthread_mutex_lock(&mu2); totalReceived += 1; pthread_mutex_unlock(&mu2)
                        }
                    }
                }
            }

            let sendThreads = senders.enumerated().map { (s, ch) in
                spawnPthread {
                    _ = try! ch.waitForRecv(count: n, timeout: .seconds(3))
                    for j in 0..<msgPerSender {
                        let msg = Array("S\(s)M\(j)".utf8)
                        if (try? ch.send(data: msg, timeout: .seconds(5))) == true {
                            pthread_mutex_lock(&mu2); totalSent += 1; pthread_mutex_unlock(&mu2)
                        }
                    }
                    ch.disconnect()
                }
            }

            for t in sendThreads { await joinThread(t) }
            for t in recvThreads { await joinThread(t) }
            pthread_mutex_destroy(&mu2)
            for r in receivers { r.disconnect() }
            ctrl.disconnect()

            #expect(totalSent == totalMsgs)
            #expect(totalReceived == totalMsgs * n)
        }
    }

    // Port of channel_nv1_broadcast — N senders, 1 receiver
    @Test("channel Nv1 — 2 and 4 senders, 1 receiver")
    func channelNv1Broadcast() async throws {
        for numSenders in [2, 4] {
            let name = "swift_stress_cnv1_\(UInt32.random(in: 0..<UInt32.max))"
            defer { Task { await Channel.clearStorage(name: name) } }

            let msgPerSender = 100
            let totalMsgs = numSenders * msgPerSender

            let receiver = Channel.connectBlocking(name: name, mode: .receiver)
            let senders  = (0..<numSenders).map { _ in Channel.connectBlocking(name: name, mode: .sender) }

            nonisolated(unsafe) var totalSent     = 0
            nonisolated(unsafe) var totalReceived = 0

            let recvThread = spawnPthread {
                for _ in 0..<totalMsgs {
                    let buf = try! receiver.recv(timeout: .seconds(5))
                    if !buf.isEmpty { totalReceived += 1 }
                }
            }

            let sendThreads = senders.enumerated().map { (s, ch) in
                spawnPthread {
                    _ = try! ch.waitForRecv(count: 1, timeout: .seconds(3))
                    for j in 0..<msgPerSender {
                        let msg = Array("S\(s)M\(j)".utf8)
                        if (try? ch.send(data: msg, timeout: .seconds(5))) == true {
                            totalSent += 1
                        }
                    }
                    ch.disconnect()
                }
            }

            for t in sendThreads { await joinThread(t) }
            await joinThread(recvThread)
            receiver.disconnect()

            #expect(totalSent == totalMsgs)
            #expect(totalReceived == totalMsgs)
        }
    }

    // Port of channel_rapid_reconnect — 20 connect/send/recv/disconnect cycles
    @Test("rapid reconnect — 20 cycles")
    func channelRapidReconnect() async throws {
        let name = "swift_stress_reconnect_\(UInt32.random(in: 0..<UInt32.max))"
        defer { Task { await Channel.clearStorage(name: name) } }

        for i in 0..<20 {
            let sender   = Channel.connectBlocking(name: name, mode: .sender)
            let receiver = Channel.connectBlocking(name: name, mode: .receiver)

            _ = try sender.waitForRecv(count: 1, timeout: .seconds(1))
            let msg = Array("iter\(i)".utf8)
            _ = try sender.send(data: msg, timeout: .seconds(1))

            let buf = try receiver.recv(timeout: .seconds(1))
            #expect(!buf.isEmpty)
            #expect(buf.bytes == msg)

            sender.disconnect(); receiver.disconnect()
        }
    }

    // Port of route_large_messages_full_range — 128B to 16KB
    @Test("route large messages — 128B to 16KB range")
    func routeLargeMessageFullRange() async throws {
        let name = "swift_stress_large_range_\(UInt32.random(in: 0..<UInt32.max))"
        defer { Task { await Route.clearStorage(name: name) } }

        // Sizes up to 4KB — 8KB+ trigger a sendLarge path that blocks the
        // cooperative thread pool when called from a pthread context.
        let sizes = [128, 256, 512, 1024, 2048, 4096]
        await Route.clearStorage(name: name)
        let sender   = try await Route.connect(name: name, mode: .sender)
        let receiver = try await Route.connect(name: name, mode: .receiver)

        final class State: @unchecked Sendable { var errors: [String] = [] }
        let state = State()

        let recvThread = spawnPthread {
            for (i, expectedSize) in sizes.enumerated() {
                let buf = try! receiver.recv(timeout: .seconds(10))
                if buf.bytes.count != expectedSize {
                    state.errors.append("msg \(i): expected \(expectedSize) got \(buf.bytes.count)")
                } else {
                    let fill = UInt8(i & 0xFF)
                    if !buf.bytes.allSatisfy({ $0 == fill }) {
                        let bad = buf.bytes.enumerated().first(where: { $0.element != fill })
                        state.errors.append("msg \(i) size=\(expectedSize): corrupt at \(bad!.offset) got \(bad!.element) want \(fill)")
                    }
                }
            }
        }
        let sendThread = spawnPthread {
            _ = try! sender.waitForRecv(count: 1, timeout: .seconds(2))
            for (i, sz) in sizes.enumerated() {
                let msg = [UInt8](repeating: UInt8(i), count: sz)
                _ = try! sender.send(data: msg, timeout: .seconds(10))
            }
        }

        await joinThread(sendThread)
        await joinThread(recvThread)
        sender.disconnect(); receiver.disconnect()
        #expect(state.errors.isEmpty, "errors: \(state.errors)")
    }

    // Port of route_large_messages_stress — 20 × 1KB messages
    @Test("route large messages stress — 20 × 1KB")
    func routeLargeMessagesStress() async throws {
        let name = "swift_stress_large_stress_\(UInt32.random(in: 0..<UInt32.max))"
        defer { Task { await Route.clearStorage(name: name) } }

        let msgCount = 20
        let msgSize  = 1024
        await Route.clearStorage(name: name)
        let sender   = try await Route.connect(name: name, mode: .sender)
        let receiver = try await Route.connect(name: name, mode: .receiver)

        final class State: @unchecked Sendable { var errors: [String] = [] }
        let state = State()

        let recvThread = spawnPthread {
            for i in 0..<msgCount {
                let buf = try! receiver.recv(timeout: .seconds(10))
                if buf.bytes.count != msgSize {
                    state.errors.append("msg \(i): wrong size \(buf.bytes.count)")
                } else {
                    let fill = UInt8(i & 0xFF)
                    if !buf.bytes.allSatisfy({ $0 == fill }) {
                        state.errors.append("msg \(i): corrupt")
                    }
                }
            }
        }
        let sendThread = spawnPthread {
            _ = try! sender.waitForRecv(count: 1, timeout: .seconds(2))
            for i in 0..<msgCount {
                let msg = [UInt8](repeating: UInt8(i), count: msgSize)
                _ = try! sender.send(data: msg, timeout: .seconds(10))
            }
        }

        await joinThread(sendThread)
        await joinThread(recvThread)
        sender.disconnect(); receiver.disconnect()
        #expect(state.errors.isEmpty, "errors: \(state.errors)")
    }
}
